CIE-5604 - configuring Edge operator (#4767)
SoftwareAG/c8y-docs: content/edge/installing-edge-bundle/install-edge-with-edge-operator.md

For more information about the structure and configuration options available in the Edge CR, see [Edge custom resource](/edge/edge-custom-resource-definition/).

### Configuring the Edge operator with a proxy and trusted TLS certificates

You can configure the Edge operator to:

>   - Route outbound traffic through a proxy server when deployed behind a proxy

Reviewer: ```suggestion
  - Route outbound traffic through a proxy server when deployed behind a proxy.
```

>   - Trust additional TLS certificates for external endpoints

Reviewer: ```suggestion
  - Trust additional TLS certificates for external endpoints.
```

You can configure the Edge operator to:
  - route outbound internet traffic through a proxy server when it is deployed behind a proxy
  - trust the TLS/SSL certificates presented by external endpoints

To configure proxy settings and trusted certificates, create or update a ConfigMap in the `c8yedge` namespace (or the namespace where Edge is deployed) with the required configuration keys described below:
  - `http_proxy` - URL of the HTTP proxy server
  - `https_proxy` - URL of the HTTPS proxy server
  - `socks_proxy` - URL of the SOCKS proxy server
  - `no_proxy` - Comma-separated list of domain suffixes, IP addresses, or CIDR ranges that should bypass the proxy. This list must include:
      - {{< management-tenant >}} and the Edge tenant domain names
      - Kubernetes Pod CIDR (Cluster pod IP address range)
      - Kubernetes Service CIDR (Cluster service IP address range)
      - Any additional domains, hosts or IP addresses that should bypass the proxy
      Example: `127.0.0.1,::1,localhost,.svc,.cluster.local,cumulocity,<edge domain names, e.g. management-myown.iot.com,myown.iot.com>,<kubernetes cluster IP range, e.g. 10.43.0.0/16>`

>   - `ca.crt` - One or more certificates in PEM format that the Edge operator and the Edge should trust in addition to the certificates already included in the system trust store. Multiple certificates can be provided by concatenating them into a single PEM bundle.

Reviewer 2: I'm wondering if using 'certificates' instead of 'certificate authority' is ambiguous here, but I'm not 100% familiar with all of the x509 terminology. Can you take an actual certificate used by a service and put it here, or ca.crt have to be something used to sign?

Also

```suggestion
  - `ca.crt` - One or more certificates in PEM format that the Edge operator and the Edge should trust in addition to publicly known certificate authorities. Multiple certificates can be provided by concatenating them into a single PEM bundle.
```
because the original implies that anything that has been added by hand to the _operating system_ trust store will automatically be trusted by the operator. This is especially risky in the private registry case, because somebody would add the CA for their private registry to the OS trust store, and then assume they don't need to add the same cert for the operator.

Author: > I'm wondering if using 'certificates' instead of 'certificate authority' is ambiguous here, but I'm not 100% familiar with all of the x509 terminology. Can you take an actual certificate used by a service and put it here, or ca.crt have to be something used to sign?

You can indeed trust either the server/leaf certificate (actual certificate used by the server) or any of the signing certificates (root or intermediary) in the trust chain. `root --> intermediary --> server certificate` 

> because the original implies that anything that has been added by hand to the _operating system_ trust store will automatically be trusted by the operator. This is especially risky in the private registry case, because somebody would add the CA for their private registry to the OS trust store, and then assume they don't need to add the same cert for the operator.

For the private registry to work, they are expected do both. I mean add the server certificate (or one of its chain) to the system trust store (in the node for the `containerd` to trust when it pulls the docker images) and also to the operator (via this ConfigMap) for it to trust while pulling the artifacts (helm charts and ORAS files) from the private registry. 
Adding the same to the JKS truststore passed to the Cumulocity core's JVM, could be our workaround if the core/platform team doesn't agree on providing a solution from their end.

You are right, I changed the text as you suggested --> `in addition to publicly known certificate authorities`  https://github.com/Cumulocity-IoT/c8y-docs/pull/4767/commits/afaa6f160bedacbf53565db21b7f50317943cb93

The following example shows a ConfigMap with proxy settings and trusted certificates:

```yaml
##
## Optional ConfigMap used to configure:
##   - Proxy settings for accessing external endpoints
##   - Additional trusted TLS/SSL certificates
##

apiVersion: v1
kind: ConfigMap
metadata:
  # Name of the ConfigMap
  name: c8yedge-operator-config
  
  # Namespace where the Edge operator is installed
  namespace: c8yedge
data:
  http_proxy: <HTTP Proxy URL>
  https_proxy: <HTTPS Proxy URL>
  socks_proxy: <SOCKS Proxy URL>

  # Comma-separated list of domain suffixes, IP addresses, or CIDR ranges that should bypass the proxy
  no_proxy: 127.0.0.1,::1,localhost,.svc,.cluster.local,cumulocity,<edge domain names, e.g. management-myown.iot.com,myown.iot.com>,<kubernetes cluster IP range, e.g. 10.43.0.0/16>

  # Trusted TLS/SSL certificates in PEM format
  ca.crt: |
    <CA_CERTIFICATES_TO_TRUST>
```

> After creating or updating the ConfigMap, configure the Edge operator to use it by setting the `operatorConfigCMName` Helm value during installation or upgrade:

Reviewer 2: I'm on the fence about this, but thinking out loud:

I wonder if it's better to put a much less verbose command here, without tediously restating the known values.
e.g.
`helm upgrade ...   --reuse-values --set operatorConfigCMName=...`

That said, I'm not sure if we can avoid `--version`, and that's a problem because `--version=2026` might caused some unnecessary operator bouncing if they're not already on the latest.

Also remove the `login`, because that's already implicit in the fact that they installed it, I think?

Author: Actually I contemplated on moving the section **Configuring the Edge operator with a proxy and trusted TLS/SSL certificates**  before **Installing the Edge operator**. But this also has a caveat, that these configurations can be created or updated post the operator installation. 
In fact, this is the specific reason we had initially decided to keep the ConfigMap name for the operator config as a predefined name. In we keep the name fixed (not configurable), the user just needs to restart the operator after creating/updating the ConfgMap.

Author: As discussed, the `custom-environment-variables` CM is renamed to `c8yedge-operator-config` and the doc is updated accordingly. The CM name is not configurable and user just has to create or update the CM with a pre-defined name `c8yedge-operator-config`, which the operator reads and updates its configurations accordingly when it is restarted.
